### ISSUE TITLE
a quick question about 'Designing the Second Toolbox - First Attempt'

### ISSUE THREAD

**Issue by fanision** [ayaanhossain/nrpcalc#3]
Hi there, thx so much for providing us with such a helpful and insightful tool and also this clear jupyter tutorial. 
My question is just when you design the **first toolbox** why did not you use a **global mode function** as you did in the **second toolbox**? Look forward to your reply.

**Comment by ayaanhossain** [comment]
Hello @fanision, thanks so much for checking out the tool. I am glad you find the tool helpful!

The first toolbox design is basically an introduction to the Maker mode and Finder mode APIs. As such, the example of the promoter toolbox is kept very simple (sequence constraint only, no structures, and no model functions are used). Of course, that design is probably not very practical, so we show how to design more realistic promoters, optimizing them for characterization experiments and high functionality. All the examples become increasingly more complex and are arranged to show different aspects of non-repetitive genetic part design. Hope this answers your questions!

**Comment by ayaanhossain** [comment]
Hi @fanision, thanks so much for checking the tool out. I am glad you find it helpful!

The first promoter toolbox is just an introductory example of what the Maker algorithm can do. It just shows how to define a sequence constraint, an easy local model function (prevent palindromic hexamers), and finally shows the practice of writing verification functions (to check if all designed promoters indeed have no palindromic hexamers inside them). The second promoter toolbox example shows how to use multiple local and global objectives by wrapping them up in a meta local or meta global model function.

At the end of the first toolbox example, when we used the verification function, it would have been better to pass it as a global model function -- you are absolutely right! This is why in the notebook we say the following

> Every promoter has passed our verification so our design objective for the first toolbox was met. As we will see in the next section, the evaluation function final_cutsite_check could have been specified to Maker directly via global_model_fn parameter, which would automatically execute the evaluation on a part, after it was completely designed, and accept/reject it accordingly.

> The benefit of passing this check as a global model function to Maker is that the algorithm can adjust the number of trials it needs depending on an auto-estimated probability of evaluation failure.

When I was writing the notebook, I didn't want to dump all the concepts in the first example. So, I showed a simple verification function and explained that it would be better to pass this as a global model function. It would then force the algorithm to continue making parts until the target toolbox size was reached. But, the first example had just one easy model function, and the final checker was to verify if everything went according to plan -- but yes, for practical use you would obviously want to specify that function as a global model function as done for the second toolbox.
